### PR TITLE
ARROW-8618: [C++] Clean up some redundant std::move()s

### PR DESCRIPTION
**Pull request by bkietz**
ARROW_ASSIGN_OR_RAISE and friends move their RHS, so calling move on those is unnecessary.

### CODE REVIEW

File: cpp/src/arrow/testing/gtest_util.h [hunk 1]
```diff
@@ -108,9 +108,9 @@
   } while (false);
 
 #define ASSIGN_OR_HANDLE_ERROR_IMPL(handle_error, status_name, lhs, rexpr) \
-  auto status_name = (rexpr);                                              \
+  auto&& status_name = (rexpr);                                            \
   handle_error(status_name.status());                                      \
-  lhs = std::move(status_name).ValueOrDie();
+  lhs = std::move(status_name).ValueUnsafe();
```
Comment (pitrou) [inline cpp/src/arrow/testing/gtest_util.h:113]: Isn't this a problem with `EXPECT_OK_AND_ASSIGN`? Basically `lhs` will get some undetermined value.

Comment (bkietz) [reply cpp/src/arrow/testing/gtest_util.h:113]: Good point. I'll change this one back to ValueOrDie

**Comment by bkietz** [comment]
https://travis-ci.org/github/bkietz/arrow/builds/730424920

**Review by pitrou — APPROVED** [review]
+1

**Pull request closed without merging**